src/commonMain/kotlin/com/configcat/Evaluator.kt

```diff
@@ -1,437 +1,1171 @@
+                    "invalid ($reason)"
+            )
+        }
+    }
 
-    init {
-        entries.appendLine("Evaluating '$key'")
+    private fun getUserAttributeForDate(
+        userCondition: UserCondition,
+        context: EvaluationContext,
+        comparisonAttribute: String,
+        userValue: Any
+    ): Double {
+        try {
+            if (userValue is DateTime) {
+                return userValue.unixMillisDouble / 1000
+            }
+            if (userValue is DateTimeTz) {
+                return userValue.local.unixMillisDouble / 1000
+            }
+            return userAttributeToDouble(userValue)
+        } catch (e: NumberFormatException) {
+            val reason =
+                "'$userValue' is not a valid Unix timestamp (number of seconds elapsed since Unix epoch)"
+            logger.warning(
+                3004,
+                ConfigCatLogMessages.getUserAttributeInvalid(
+                    context.key,
+                    userCondition,
+                    reason,
+                    comparisonAttribute
+                )
+            )
+            throw RolloutEvaluatorException(
+                "cannot evaluate, the User.$comparisonAttribute attribute is " +
+                    "invalid ($reason)"
+            )
+        }
     }
 
-    fun logReturnValue(value: Any) {
-        entries.appendLine("Returning $value")
+    private fun getUserAttributeAsString(
+        key: String,
+        userCondition: UserCondition,
+        userAttributeName: String,
+        userValue: Any
+    ): String {
+        if (userValue is String) {
+            return userValue
+        }
+        val userAttributeToString = userAttributeToString(userValue) ?: ""
+        logger.warning(
+            3005,
+            ConfigCatLogMessages.getUserObjectAttributeIsAutoConverted(
+                key,
+                userCondition,
+                userAttributeName,
+                userAttributeToString
+            )
+        )
+
+        return userAttributeToString
     }
 
-    fun logPercentageEvaluationReturnValue(value: Any) {
-        entries.appendLine("Evaluating % options. Returning $value.")
+    private fun userAttributeToString(userValue: Any?): String? {
+        if (userValue == null) {
+            return null
+        }
+        if (userValue is String) {
+            return userValue
+        }
+        if (userValue is Array<*> && userValue.all { it is String }) {
+            return Constants.json.encodeToString(userValue as Array<String>)
+        }
+        if (userValue is List<*> && userValue.all { it is String }) {
+            return Constants.json.encodeToString(userValue as List<String>)
+        }
+        if (userValue is Float) {
+            // toSting and toDouble hack needed in case of float. else the 0.12345 is converted to something like 0.1234500120130
+            return doubleToString(userValue.toString().toDouble())
```

Comment:
I checked the reference implementation and it also does the "0.12345 is converted to something like 0.1234500120130" thing.

This is how float to double conversion works according to the IEEE 754 standard, so let's say that this is by design and remove the hack.